src/typechecker.erl

```diff
@@ -2890,6 +2906,106 @@ check_clause(_Env, _ArgsTy, _ResTy, Term) ->
     io:format("DEBUG: check_clause term: ~p~n", [Term]),
     throw(check_clause).
 
+%% Refine types by matching clause. MatchedTys are the types exhausted by
+%% each pattern in the previous clause.
+refine_clause_arg_tys(Tys, MatchedTys, [], TEnv) ->
+    Ty        = type(tuple, Tys),
+    MatchedTy = type(tuple, MatchedTys),
+    try refine(Ty, MatchedTy, TEnv) of
+        ?type(tuple, RefTys) ->
+            RefTys;
+        ?type(none) ->
+            Tys; %% We could return [none(), ...] here if we'd want to do
+                 %% exhaustiveness checking
+        ?type(union, _) ->
+            Tys %% Multiple possibilities => don't refine
+    catch
+        no_refinement -> Tys
+    end;
+refine_clause_arg_tys(Tys, _TysBounds, _Guards, _TEnv) ->
+    Tys.
+
+%% Normalize, refine, revert normalize if no refinement.
+%% May throw no_refinement.
+refine(OrigTy, Ty, TEnv) ->
+    NormTy = normalize(OrigTy, TEnv),
+    case refine_ty(NormTy, normalize(Ty, TEnv), TEnv) of
+        NormTy -> OrigTy;
+        RefTy  -> RefTy
+    end.
+
+%% May throw no_refinement.
+refine_ty(?type(union, Tys), Ty, TEnv) ->
+    normalize(type(union, [refine(U, Ty, TEnv) || U <- Tys]), TEnv);
+refine_ty(?type(tuple, Tys1), ?type(tuple, Tys2), TEnv)
+  when length(Tys1) > 0 ->
+    %% Non-empty tuple
+    RefTys = [refine(Ty1, Ty2, TEnv) || {Ty1, Ty2} <- lists:zip(Tys1, Tys2)],
+    %% {a|b, a|b} \ {a,a} => {a|b, b} | {b, a|b}
+    PossibleTyss = [case RefTy of
+                        Ty          -> [Ty];
+                        ?type(none) -> [Ty];
+                        _           -> [Ty, RefTy]
+                    end || {Ty, RefTy} <- lists:zip(Tys1, RefTys)],
+    [_|TuplesElems] = pick_from_each(PossibleTyss),
+    Tuples = [type(tuple, TupleElems) || TupleElems <- TuplesElems],
+    normalize(type(union, Tuples), TEnv);
+refine_ty(?type(any), _Ty, _) ->
+    throw(no_refinement);
```

Comment:
You may be right. I have to think on this.

Reply:
I have done a bit more thinking (but maybe not enough). I believe the problem is the `Ty` case in refinement of tuples:
```erlang
    PossibleTyss = [case RefTy of
                        Ty          -> [Ty];
                        ?type(none) -> [Ty];
                        _           -> [Ty, RefTy]
                    end || {Ty, RefTy} <- lists:zip(Tys1, RefTys)],
```
Note that this is the same as the case for `none()`. That is, no refinement is treated the same as complete covering, which surely can't be right. I think this case should simply be removed:
```erlang
    PossibleTyss = [case RefTy of
                        ?type(none) -> [Ty];
                        _           -> [Ty, RefTy]
                    end || {Ty, RefTy} <- lists:zip(Tys1, RefTys)],
```
There really shouldn't have to be a special case for a no-op refinement, and needing one is a huge red flag that there's something iffy (what if the refined type is semantically the same but not syntactically?). With this change, and the change to `pick_from_each`, I believe imprecise refinements can be handled by simply returning the original type.

Reply:
That seems right. But if `Ty == RefTy` for one of the types, say for example `RefTy2==Ty2`, and in the result union exactly one Ty replaced by RefTy so that in this example, the tuple `{Ty1, RefTy2, Ty3, ...}` is included in the result union, the result is the original tuple. This means there is no refinement. Therefore, we can as well have the `throw(no_refinement)` in various places, to abort refinement.

Reply:
(I wanted to simplify `[Ty, Ty]` to `[Ty]` just to limit the 2^N explosion when possible, but now it won't be 2^N anyway.)

Reply:
I think removing the clause which prevents the refinement `any() \ any() => none()` may solve your `fails` example above. The second `any()` doesn't actually mean *unknown* in this case. In this case, it's just `Ty1 == Ty2` (which means that the type is exhausted by the pattern).

Reply:
Yep, it works.